src/pg_copy.c

```diff
@@ -0,0 +1,552 @@
+{
+    StringInfo buf = makeStringInfo();
+    ListCell* cell;
+    char sep;
+    char const* qualifiedName = quote_qualified_identifier(copyStatement->relation->schemaname,
+                                                           copyStatement->relation->relname);
+    appendStringInfo(buf, "COPY %s_%ld ", qualifiedName, (long)shardId);    
+    if (copyStatement->attlist) { 
+        sep = '(';
+        foreach(cell, copyStatement->attlist)
+        {
+            appendStringInfoChar(buf, sep);
+            appendStringInfoString(buf, strVal(lfirst(cell)));
+            sep = ',';
+        }
+        appendStringInfoChar(buf, ')');        
+    }
+    appendStringInfoString(buf, "FROM STDIN");
+
+    if (copyStatement->options) { 
+        appendStringInfoString(buf, " WITH ");
+        sep = '(';
+        foreach(cell, copyStatement->options)
+        {
+            DefElem* def = (DefElem *) lfirst(cell);     
+            appendStringInfo(buf, "%c%s '%s'", sep, def->defname, defGetString(def));
+            sep = ',';
+        }
+        appendStringInfoChar(buf, ')');        
+    }   
+    return buf->data;
+}
+
+static bool PgCopyPrepareTransaction(ShardId shardId, PGconn* conn, void* arg, bool status)
+{
+
+	char const* relationName = (char*)arg;
+	PgShardTransactionManager const* tmgr = &PgShardTransManagerImpl[PgShardCurrTransManager];
+	PQputCopyEnd(conn, NULL);
+	return tmgr->Prepare(conn, relationName, shardId);
+}
+
+static bool PgCopyAbortTransaction(ShardId shardId, PGconn* conn, void* arg, bool status)
+{
+	char const* relationName = (char*)arg;
+	PgShardTransactionManager const* tmgr = &PgShardTransManagerImpl[PgShardCurrTransManager];
+
+	if (status) 
+	{
+		tmgr->RollbackPrepared(conn, relationName, shardId);
+	} else {
+		PQputCopyEnd(conn, "Aborted because of failure on some shard");
+		tmgr->Rollback(conn);
+	}
+	PQfinish(conn);
+	return true;
+}
+
+static bool PgCopyEndTransaction(ShardId shardId, PGconn* conn, void* arg, bool status)
+{
+	char const* relationName = (char*)arg;
+	PgShardTransactionManager const* tmgr = &PgShardTransManagerImpl[PgShardCurrTransManager];
+	
+	if (status)
+	{
+		tmgr->CommitPrepared(conn, relationName, shardId);
+	} else {
+		PQputCopyEnd(conn, "Aborted because of failure on some shard");
+		tmgr->Rollback(conn);
+	}
+	PQfinish(conn);
+	return true;
+}
+	
+/*
+ * Handle copy to/from distributed table
+ */
+bool PgShardCopy(CopyStmt *copyStatement, char const* query)
+{		 
+	RangeVar *relation = copyStatement->relation;
+	ListCell *shardIntervalCell = NULL;
+	ListCell *taskPlacementCell = NULL;
+	List *shardIntervalList = NULL;
+	char *relationName = NULL;
+	PgShardTransactionManager const* tmgr = &PgShardTransManagerImpl[PgShardCurrTransManager];
+	bool failOK = true;
+	Oid tableId = 0;
+
+	if (relation == NULL)
+	{
+		return false;
+	}
+	tableId = RangeVarGetRelid(relation, NoLock, failOK);
+	if (IsDistributedTable(tableId))
+	{
```

Comment:
could you move this code block into a separate function? We really need to get this function under 90 columns.